akka-remote/src/main/scala/akka/remote/artery/FlightRecorder.scala

```diff
@@ -0,0 +1,360 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.{ ByteBuffer, ByteOrder }
+import java.nio.channels.FileChannel
+import java.nio.file.StandardOpenOption
+import java.util.concurrent.atomic.AtomicBoolean
+
+import akka.util.ByteString
+import org.agrona.BitUtil
+import org.agrona.concurrent.MappedResizeableBuffer
+
+import scala.annotation.tailrec
+
+/**
+ * INTERNAL API
+ */
+private[remote] trait EventSink {
+  def alert(code: Int, metadata: Array[Byte]): Unit
+  def loFreq(code: Int, metadata: Array[Byte]): Unit
+  def hiFreq(code: Long, param: Long): Unit
+
+  def flushHiFreqBatch(): Unit
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] object IgnoreEventSink extends EventSink {
+  override def alert(code: Int, metadata: Array[Byte]): Unit = ()
+  override def loFreq(code: Int, metadata: Array[Byte]): Unit = ()
+  override def flushHiFreqBatch(): Unit = ()
+  override def hiFreq(code: Long, param: Long): Unit = ()
+}
+
+/**
+ * INTERNAL API
+ *
+ * Update clock at various resolutions and aquire the resulting timestamp.
+ */
+private[remote] trait EventClock {
+
+  def updateWallClock(): Unit
+  def updateHighSpeedClock(): Unit
+
+  def getWallClockPart: Long
+  def getHighSpeedPart: Long
+
+}
+
+/**
+ * INTERNAL API
+ *
+ * This class is not thread-safe
+ */
+private[remote] class EventClockImpl extends EventClock {
+
+  private[this] var wallClock: Long = 0
+  private[this] var highSpeedClock: Long = 0
+  private[this] var highSpeedClockOffset: Long = 0
+
+  updateWallClock()
+
+  override def updateWallClock(): Unit = {
+    wallClock = System.currentTimeMillis()
+    highSpeedClockOffset = System.nanoTime()
+    highSpeedClock = 0
+  }
+
+  override def updateHighSpeedClock(): Unit = {
+    // TODO: Update wall clock once in a while
+    highSpeedClock = System.nanoTime() - highSpeedClockOffset
+  }
+
+  override def getWallClockPart: Long = wallClock
+  override def getHighSpeedPart: Long = highSpeedClock
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] object RollingEventLogSection {
+  val HeadPointerOffset = 0L
+  val LogStateOffset = 8L
+  val RecordsOffset = 16L
+  val LogOffset = 0L
+
+  // Log states
+  val Empty = 0
+  val Live = 1
+  val Snapshot = 2
+
+  // Slot states
+  val Committed = 0
+  val Dirty = 1
+
+  val CommitEntrySize = 4
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] class RollingEventLogSection(
+  fileChannel:   FileChannel,
+  offset:        Long,
+  entryCount:    Long,
+  logBufferSize: Long,
+  recordSize:    Int) extends AtomicBoolean {
```

Comment:
Why is a `RollingEventLogSection` an `AtomicBoolean` non obvious is-a-relation, add a comment to explain?